ubuntu/main.cc

```diff
@@ -0,0 +1,257 @@
+  FlutterWindowMetricsEvent event = {};
+  event.struct_size = sizeof(event);
+  event.width = width;
+  event.height = height;
+  event.pixel_ratio = 1.0;
+  FlutterEngineSendWindowMetricsEvent(
+      reinterpret_cast<FlutterEngine>(glfwGetWindowUserPointer(window)),
+      &event);
+}
+
+static bool GLFWMakeContextCurrent(void *user_data) {
+  GLFWwindow *window = reinterpret_cast<GLFWwindow *>(user_data);
+  glfwMakeContextCurrent(window);
+  return true;
+}
+
+static bool GLFWClearContext(void *user_data) {
+  glfwMakeContextCurrent(nullptr);
+  return true;
+}
+
+static bool GLFWPresent(void *user_data) {
+  GLFWwindow *window = reinterpret_cast<GLFWwindow *>(user_data);
+  glfwSwapBuffers(window);
+  return true;
+}
+
+static uint32_t GLFWGetActiveFbo(void *user_data) { return 0; }
+
+// Clears the GLFW window to Material Blue-Grey.
+//
+// This function is primarily to fix an issue when the Flutter Engine is
+// spinning up, wherein artifacts of existing windows are rendered onto the
+// canvas for a few moments.
+//
+// This function isn't necessary, but makes starting the window much easier on
+// the eyes.
+static void GLFWClearCanvas(GLFWwindow *window) {
+  glfwMakeContextCurrent(window);
+  // This color is Material Blue Grey.
+  glClearColor(236, 239, 241, 0);
+  glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
+  glFlush();
+  glfwSwapBuffers(window);
+  glfwMakeContextCurrent(nullptr);
+}
+
+static bool RunFlutterEngine(GLFWwindow *window,
+                             const std::string &flutter_app_directory,
+                             const std::string &main_path,
+                             const std::string &icu_data_path) {
+  FlutterRendererConfig config = {};
+  config.type = kOpenGL;
+  config.open_gl.struct_size = sizeof(config.open_gl);
+  config.open_gl.make_current = GLFWMakeContextCurrent;
+  config.open_gl.clear_current = GLFWClearContext;
+  config.open_gl.present = GLFWPresent;
+  config.open_gl.fbo_callback = GLFWGetActiveFbo;
+  std::string assets_path = flutter_app_directory + "/build/flutter_assets";
+  std::string full_main_path = flutter_app_directory + "/" + main_path;
+  std::string packages_path = flutter_app_directory + "/.packages";
+  FlutterProjectArgs args = {};
+  args.struct_size = sizeof(FlutterProjectArgs);
+  args.assets_path = assets_path.c_str();
+  args.main_path =
+      main_path.empty() ? main_path.c_str() : full_main_path.c_str();
+  // Packages path is ignored if main_path is empty, so this can be added in
+  // as-is.
+  args.packages_path = packages_path.c_str();
+  args.icu_data_path = icu_data_path.c_str();
+  FlutterEngine engine = nullptr;
+  auto result =
+      FlutterEngineRun(FLUTTER_ENGINE_VERSION, &config, &args, window, &engine);
+  if (result != kSuccess || engine == nullptr) {
+    return false;
+  }
+  GLFWClearCanvas(window);
+  glfwSetWindowUserPointer(window, engine);
+  GLFWwindowSizeCallback(window, kInitialWindowWidth, kInitialWindowHeight);
```

Comment:
You'll want to use glfwGetWindowSize here once this moves into a separate file (see below).

Reply:
Done!